bindings/rust/s2n-tls-tokio/src/contract.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use core::{
+    sync::atomic::{AtomicBool, Ordering},
+    task::{Context, Poll, Waker},
+};
+use std::{sync::Arc, task::Wake};
+
+/// Checks that if a function returns [`Poll::Pending`], then the function called [`Waker::clone`],
+/// [`Waker::wake`], or [`Waker::wake_by_ref`] on the [`Context`]'s [`Waker`].
+pub struct Contract {
+    state: Arc<State>,
+    waker: Waker,
+}
+
+struct State {
+    inner: Waker,
+    wake_called: AtomicBool,
+}
+
+impl Wake for State {
+    #[inline]
+    fn wake(self: Arc<Self>) {
+        Wake::wake_by_ref(&self)
+    }
+
+    #[inline]
+    fn wake_by_ref(self: &Arc<Self>) {
+        self.wake_called.store(true, Ordering::Release);
+        self.inner.wake_by_ref();
+    }
+}
+
+impl Contract {
+    /// Wraps a [`Context`] in the contract checker
+    #[inline]
+    pub fn new(cx: &mut Context) -> Self {
+        let state = State {
+            inner: cx.waker().clone(),
+            wake_called: AtomicBool::new(false),
+        };
+        let state = Arc::new(state);
+        let waker = Waker::from(state.clone());
+        Self { state, waker }
+    }
+
+    /// Returns a new [`Context`] to be checked
+    #[inline]
+    pub fn context(&self) -> Context {
+        Context::from_waker(&self.waker)
+    }
+
+    /// Checks the state of the waker based on the provided `outcome`
+    #[inline]
+    #[track_caller]
+    pub fn check_outcome<T>(self, outcome: &Poll<T>) {
```

Comment:
Maybe "assert_outcome" to make it clearer that it's an assertion and may panic?

Reply:
This code has already been reviewed in s2n-quic in https://github.com/aws/s2n-quic/pull/2248. I don't want to diverge the two, as it will make it hard to merge in the future.